`gh pr edit`: Only fetch org teams for reviewers when required

BagToad opened the pull request:
### Description

This pull request refactors how pull request reviewer updates are handled, switching from the GraphQL API to the REST API for adding and removing reviewers. It also improves the logic for managing reviewer lists, especially in non-interactive and interactive editing modes, and updates the test suite to cover these changes and edge cases.

Also fixes a bug where the PR author was included in the interactive list of reviewers. With GraphQL this selection would be ignored, but the REST API actually returns an error if the PR author is provided, so we need to fix that now to avoid confusion.

fixes #4844

### Considerations

Moving to the REST API with this allows us to add/remove as separate operations, eliminating the need to fetch org teams and org team IDs. This is good, because it allows the `GITHUB_TOKEN` and other repo scoped authentication types to proceed with requesting a user review.

However, there's a drawback that we now will require _two operations_, one for adding reviewers and one for removing. 

I am fine with this trade-off because I suspect the typical use-case will only involve either requesting a reviewer or removing a reviewer as discrete operations. Meaning, in most cases we'll never actually perform two API calls. Nonetheless the risk is there.

This implementation also proposes that these two API requests be performed concurrently. I suspect this is optimal, but I am unsure of any sort of race-condition that could arise. In particular in the case of removing and adding the same reviewer, but that could be seen as user error. 

### Testing

This was tested in Actions to prove that it will allow requesting a user reviewer without requiring access to teams, meaning the `GITHUB_TOKEN` could be used where currently `gh` will fail.

<details><summary>Test workflow</summary>
<p>

```Yaml

name: Test Team Reviewer

on:
  workflow_dispatch:
    inputs:
      pr_number:
        description: 'PR number to add reviewer to'
        required: true
        type: string

jobs:
  add-individual-reviewer-local-gh:
    runs-on: self-hosted
    continue-on-error: true
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Add individual reviewer using local gh binary (should pass if permissions allow)
        env:
            GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          <LOCAL_GH_PATH>/gh pr edit ${{ github.event.inputs.pr_number }} --add-reviewer "<INDIVIDUAL_REVIEWER_USERNAME>"

  add-team-reviewer-local-gh:
    runs-on: self-hosted
    continue-on-error: true
    needs: add-individual-reviewer-local-gh
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Add team reviewer using local gh binary (may fail if token lacks team review permission)
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          <LOCAL_GH_PATH>/gh pr edit ${{ github.event.inputs.pr_number }} --add-reviewer "<ORG_SLUG>/<TEAM_SLUG>"

  add-team-reviewer-system-gh:
    runs-on: self-hosted
    continue-on-error: true
    needs: add-team-reviewer-local-gh
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Add team reviewer using system gh (may fail if permissions insufficient)
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          gh pr edit ${{ github.event.inputs.pr_number }} --add-reviewer "<ORG_SLUG>/<TEAM_SLUG>"

  add-individual-reviewer-system-gh:
    runs-on: self-hosted
    continue-on-error: true
    needs: add-team-reviewer-system-gh
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Add individual reviewer using system gh
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          gh pr edit ${{ github.event.inputs.pr_number }} --add-reviewer "<INDIVIDUAL_REVIEWER_USERNAME>"
```

</p>
</details>  

You can see the first job is successful, which is what we're going for. We want the `GITHUB_TOKEN` to be able to request a review from a user as long as there isn't teams included in the review request.

<img width="602" height="101" alt="image" src="https://github.com/user-attachments/assets/d8d471d5-5d1a-46d0-890d-d83e75bdab9f" />

> [!NOTE]
> This was also tested with a team reviewer _already_ requested to prove that we will not fetch team reviewers beyond what the PR metadata gives us already through the non-interactive flow. Therein lies the key reason to switch to the REST API here.

### Further Improvements

It should be noted that the REST API doesn't return a satisfying or useful error message when you provide it a team reviewer without having access to org teams, as in the case of the `GITHUB_TOKEN`:

<img width="1110" height="56" alt="image" src="https://github.com/user-attachments/assets/663e48c1-09ce-4a9d-bca4-8e6ef68b0434" />

Perhaps there is further work to be done in the future to improve this error messaging client-side.

BagToad committed:
Refactor PR reviewer editing to use REST API and optimize team fetch

Switches pull request reviewer add/remove operations from GraphQL to the REST API, enabling separate add and remove calls for reviewers and teams. Refactors reviewer editing logic to avoid fetching organization teams unless required for interactive editing, improving performance for non-interactive flows. Updates tests and supporting code to reflect the new reviewer management and metadata fetching behavior.

BagToad commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -237,7 +240,7 @@ func editRun(opts *EditOptions) error {
 
 	findOptions := shared.FindOptions{
 		Selector: opts.SelectorArg,
```
I don't think that's necessary

BagToad commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -320,36 +335,71 @@ func editRun(opts *EditOptions) error {
+	var addUsers []string
+	var addTeams []string
+	for _, r := range editable.Reviewers.Value {
+		if strings.ContainsRune(r, '/') {
+			teamSlug := strings.Split(r, "/")[1]
+			addTeams = append(addTeams, teamSlug)
+		} else {
+			addUsers = append(addUsers, r)
+		}
```
hmmm I don't think I see a situation where the string contains a `/` but then doesn't have 2 parts. Unless maybe it ends with a `/`?

babakks replied:
Even if `r` ends with `/` you'd get an empty `teamSlug`, so there'll be no panic. An empty value will probably cause the API to return an error, so all good.

babakks commented on api/queries_pr.go:
```diff
@@ -629,17 +631,55 @@ func CreatePullRequest(client *Client, repo *Repository, params map[string]inter
-				ID string
-			}
-		} `graphql:"requestReviews(input: $input)"`
+// AddPullRequestReviews updates the requested reviewers on a pull request using the REST API.
```
How about highlighting the fact that this is add operation (as opposed to set/put)?


```suggestion
// AddPullRequestReviews adds the given users/teams to requested reviewers on a pull request using the REST API.
```

babakks commented on api/queries_pr.go:
```diff
@@ -629,17 +631,55 @@ func CreatePullRequest(client *Client, repo *Repository, params map[string]inter
+		return err
+	}
+	// The endpoint responds with the updated pull request object; we don't need it here.
+	return client.REST(repo.RepoHost(), "POST", path, buf, nil)
```
Not a problem for this PR, but the way we handle these `nil` args down in `go-gh` is weird.

babakks commented on api/queries_pr.go:
```diff
@@ -629,17 +631,55 @@ func CreatePullRequest(client *Client, repo *Repository, params map[string]inter
-	err := client.Mutate(repo.RepoHost(), "PullRequestUpdateRequestReviews", &mutation, variables)
-	return err
+
+	path := fmt.Sprintf("repos/%s/%s/pulls/%d/requested_reviewers", repo.RepoOwner(), repo.RepoName(), prNumber)
```
We should apply `url.PathEscape`:


```suggestion
	path := fmt.Sprintf("repos/%s/%s/pulls/%d/requested_reviewers", url.PathEscape(repo.RepoOwner()), url.PathEscape(repo.RepoName()), prNumber)
```

babakks commented on api/queries_pr.go:
```diff
@@ -629,17 +631,55 @@ func CreatePullRequest(client *Client, repo *Repository, params map[string]inter
+		teams = []string{}
+	}
+
+	path := fmt.Sprintf("repos/%s/%s/pulls/%d/requested_reviewers", repo.RepoOwner(), repo.RepoName(), prNumber)
```
Same:

```suggestion
	path := fmt.Sprintf("repos/%s/%s/pulls/%d/requested_reviewers", url.PathEscape(repo.RepoOwner()), url.PathEscape(repo.RepoName()), prNumber)
```

Kokomg1 replied:
🥰

babakks commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -298,6 +301,18 @@ func editRun(opts *EditOptions) error {
+		// as it is not a valid option for a reviewer.
+		// The REST API will return an error if we
+		// attempt to add the PR author as a reviewer.
+		// However, the GraphQL API will silently ignore it.
```
**nitpick:** Maybe collapse these lines in two?

BagToad replied:
Done in 848faf81155e72c70ffc28fb7dbb71f3157ace5c

babakks commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -320,36 +335,71 @@ func editRun(opts *EditOptions) error {
-	union := githubv4.Boolean(false)
-	reviewsRequestParams := githubv4.RequestReviewsInput{
-		PullRequestID: id,
-		Union:         &union,
-		UserIDs:       ghIds(userIds),
-		TeamIDs:       ghIds(teamIds),
+
+	// Rebuild the Value slice from non-interactive flag input.
+	if len(editable.Reviewers.Add) != 0 || len(editable.Reviewers.Remove) != 0 {
+		s := set.NewStringSet()
```
Do we really need the second `if`? I mean, isn't the condition always `true`?

BagToad replied:
Unfortunately the `.Add` and `.Remove` values are only populated non-interactively. 

Interactively, the prompter gives us the direct `.Value` with reviewers removed or added already. In other words, the prompter gives the final reviewer list only.

So, when this is false, it means it's the interactive flow, so we shouldn't rebuild the `.Value` slice since the prompter already set it.

babakks replied:
Makes sense.

babakks commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -320,36 +335,71 @@ func editRun(opts *EditOptions) error {
+		} else {
+			addUsers = append(addUsers, r)
+		}
+	}
```
**nitpick:** This piece can be extracted as a small helper function and be reused here twice:


```go
func partitionUsersAndTeams(values []string) (users []string, teams []string) {
	for _, r := range values {
		if strings.ContainsRune(r, '/') {
			teamSlug := strings.Split(r, "/")[1]
			teams = append(teams, teamSlug)
		} else {
			users = append(users, r)
		}
	}
    return users, teams
}
```

BagToad replied:
Done in 848faf81155e72c70ffc28fb7dbb71f3157ace5c

babakks commented on pkg/cmd/pr/edit/edit.go:
*Just wanted to try this new commenting on unchanged lines.*

Yeah, this is unused.

babakks commented on pkg/cmd/pr/shared/editable.go:
```diff
@@ -428,17 +396,20 @@ func FieldsToEditSurvey(p EditPrompter, editable *Editable) error {
 }
 
 func FetchOptions(client *api.Client, repo ghrepo.Interface, editable *Editable) error {
+	// Determine whether to fetch organization teams.
+	// Interactive reviewer editing (Edited true, but no Add/Remove slices) still needs
+	// team data for selection UI. For non-interactive flows, we never need to fetch teams.
+	teamReviewers := false
+	if editable.Reviewers.Edited {
+		// This is likely an interactive flow since edited is set but no mutations to
+		// Add/Remove slices, so we need to load the teams.
+		if len(editable.Reviewers.Add) == 0 && len(editable.Reviewers.Remove) == 0 {
+			teamReviewers = true
+		}
+	}
```
A bit confused by this. Correct me if I'm wrong, but I tend to think `editable.Reviewers.Edited` is only `true` if either or both of `--add-reviewer` or `--remove-reviewer` options were provided, in which case the inner `if` condition would never be `true`.


Probably the reason is I'm lost in the woods of `Editable`!

babakks replied:
I now understand the situation here. All good.

babakks commented on pkg/cmd/pr/edit/edit_test.go:
```diff
@@ -354,7 +354,7 @@ func Test_editRun(t *testing.T) {
 		name      string
 		input     *EditOptions
-		httpStubs func(*httpmock.Registry)
+		httpStubs func(*httpmock.Registry, *testing.T)
```
**nitpick:** My OCD brain is used to see `testing.T` as the first parameter.

babakks commented on pkg/cmd/pr/edit/edit_test.go:
```diff
@@ -804,10 +919,21 @@ func Test_editRun(t *testing.T) {
-	reg.Register(
-		httpmock.GraphQL(`query RepositoryAssignableActors\b`),
-		httpmock.StringResponse(`
+type mockRepoMetadataOptions struct {
+	reviewers     bool
+	teamReviewers bool // reviewers must also be true for this to have an effect.
+	assignees     bool
+	labels        bool
+	projects      bool // includes both legacy (v1) and v2
+	milestones    bool
+}
```
Thanks for this! 🙏

babakks commented on pkg/cmd/pr/edit/edit_test.go:
```diff
@@ -530,16 +542,107 @@ func Test_editRun(t *testing.T) {
+			httpStubs: func(reg *httpmock.Registry, t *testing.T) {
+				// reviewer add includes team but non-interactive Add/Remove provided -> no team fetch
+				mockRepoMetadata(reg, mockRepoMetadataOptions{reviewers: true})
+				reg.Exclude(t, httpmock.GraphQL(`query OrganizationTeamList\b`))
```
Maybe copy/paste the same comment from the previous test case:


```suggestion
				// explicitly assert that no OrganizationTeamList query occurs
				reg.Exclude(t, httpmock.GraphQL(`query OrganizationTeamList\b`))
```

babakks commented on pkg/cmd/pr/edit/edit_test.go:
```diff
@@ -530,16 +542,107 @@ func Test_editRun(t *testing.T) {
+			},
+			httpStubs: func(reg *httpmock.Registry, t *testing.T) {
+				mockRepoMetadata(reg, mockRepoMetadataOptions{reviewers: true})
+				reg.Exclude(t, httpmock.GraphQL(`query OrganizationTeamList\b`))
```
```suggestion
				// explicitly assert that no OrganizationTeamList query occurs
				reg.Exclude(t, httpmock.GraphQL(`query OrganizationTeamList\b`))
```

babakks reviewed (approved):
LGTM, bar a few necessary fixes (e.g. `url.PathEscape`).

BagToad committed:
Remove unused ghIds function and githubv4 import

Deleted the unused ghIds helper function and the associated githubv4 import from edit.go to clean up the codebase.

BagToad committed:
Escape repo owner and name in PR reviewer API paths

Updated AddPullRequestReviews and RemovePullRequestReviews to use url.PathEscape for repo owner and name in API paths. This ensures correct handling of special characters in repository identifiers.

BagToad committed:
Refactor reviewer partitioning in PR edit command

Extracted logic for splitting reviewer identifiers into users and teams into a new helper function, partitionUsersAndTeams. Updated updatePullRequestReviews to use this function for both adding and removing reviewers, improving code clarity and maintainability. Also clarified comments regarding PR author handling.

BagToad committed:
Fix typo in error message for required flags

Corrected '--tile' to '--title' in the error message shown when required flags are missing in non-interactive mode.

BagToad committed:
Remove default empty slices in RemovePullRequestReviews

Eliminates unnecessary initialization of users and teams to empty slices in RemovePullRequestReviews. Also updates the request body struct to use 'omitempty' for reviewers and team_reviewers, ensuring empty fields are omitted from the JSON payload.

babakks commented on pkg/cmd/pr/edit/edit.go:
```diff
@@ -170,7 +172,7 @@ func NewCmdEdit(f *cmdutil.Factory, runF func(*EditOptions) error) *cobra.Comman
 
 			if opts.Interactive && !opts.IO.CanPrompt() {
-				return cmdutil.FlagErrorf("--tile, --body, --reviewer, --assignee, --label, --project, or --milestone required when not running interactively")
+				return cmdutil.FlagErrorf("--title, --body, --reviewer, --assignee, --label, --project, or --milestone required when not running interactively")
```
Good catch! 👏

BagToad committed:
Fix argument order in httpStubs test functions

Swaps the argument order of the httpStubs functions in edit_test.go to match the expected (t *testing.T, reg *httpmock.Registry) signature. This improves consistency and prevents potential confusion or errors when calling these test helpers.

BagToad committed:
Apply suggestion from @babakks

Co-authored-by: Babak K. Shandiz <babakks@github.com>

BagToad committed:
Apply suggestion from @babakks

Co-authored-by: Babak K. Shandiz <babakks@github.com>

mishu0 commented:
Mishu017

mishu0 commented:
[kw/do-not-request-org-teams-for-reviewer-set](url)
